### PR TITLE
support shard level split on read path

### PR DESCRIPTION
**Pull request by penghuo**
### Description
* A partition of an OpenSearchTable is backed by an OpenSearch Index. Each partition is split into a configurable number of shards, which are then distributed across the cluster.
* Update doc, https://github.com/penghuo/penghuo-opensearch-spark/blob/issue396/docs/opensearch-table.md#inputpartition.
* Performance test track at https://github.com/opensearch-project/opensearch-spark/issues/403.

### Test
1. Test with EMR-S. The task is scheduled based on split count. For instance, index has 5 shards, then 5 tasks are scheduled.
```
24/06/29 00:24:43 INFO DAGScheduler: Got map stage job 0 (main at NativeMethodAccessorImpl.java:0) with 5 output partitions
24/06/29 00:24:43 INFO DAGScheduler: Submitting 5 missing tasks from ShuffleMapStage 0 (MapPartitionsRDD[5] at main at NativeMethodAccessorImpl.java:0) (first 15 tasks are for partitions Vector(0, 1, 2, 3, 4))
```

2. Test with AWS OpenSearch Multi-AZ with Standby domain.  it support `preference:_shard:` paramaters.

#### Single Index with 5 shards
The table below presents the p90 query latency results for both the partitioned and non-partitioned test cases. Across all queries, the results with partitioning show significantly lower times compared to the non-partitioned results.

| Query                                                                                                     | p90 ms (without partition) | p90 ms (with partition) |
|-----------------------------------------------------------------------------------------------------------|------------------------------:|----------------------------:|
| SELECT COUNT(*) FROM dev.default.`logs-181998`                                                            |                         42296 |                       17031 |
| SELECT COUNT(*) FROM dev.default.`logs-181998` WHERE status <> 0;                                         |                         44142 |                       16667 |
| SELECT COUNT(*), AVG(size) FROM dev.default.`logs-181998`;                                                |                         44584 |                       18775 |
| SELECT AVG(CAST(size AS BIGINT)) FROM dev.default.`logs-181998`;                                          |                         43575 |                       19474 |
| SELECT MIN(`@timestamp`), MAX(`@timestamp`) FROM dev.default.`logs-181998`;                               |                         43533 |                       19130 |
| SELECT status, COUNT() FROM dev.default.logs-181998 WHERE status <> 0 GROUP BY status ORDER BY COUNT() DESC; |                      43952 |                       19661 |

#### Multiple Indices, each index has 5 shards
The table below presents the p90 query latency results for both the partitioned and non-partitioned test cases when query index wildcard. Across all queries, the results with partitioning show significantly lower times compared to the non-partitioned results.

| Query                                                                                                     | p90 ms (without partition) | p90 ms (with partition) |
|-----------------------------------------------------------------------------------------------------------|------------------------------:|----------------------------:|
| SELECT COUNT(*) FROM dev.default.\`logs-1*\`                                                            |                        228049 |                       52123 |
| SELECT COUNT(*) FROM dev.default.\`logs-1*\` WHERE status <> 0;                                         |                        215548 |                       43631 |
| SELECT COUNT(*), AVG(size) FROM dev.default.\`logs-1*\`;                                                |                        222332 |                       51583 |
| SELECT AVG(CAST(size AS BIGINT)) FROM dev.default.\`logs-1*\`;                                          |                        215273 |                       51805 |
| SELECT MIN(\`@timestamp\`), MAX(\`@timestamp\`) FROM dev.default.\`logs-1*\`;                               |                        224576 |                       45914 |
| SELECT status, COUNT(*) FROM dev.default.\`logs-1*\` WHERE status <> 0 GROUP BY status ORDER BY COUNT(*) DESC; |                        189594 |                       46758 |

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/396

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java [hunk 1]
```diff
@@ -74,6 +74,16 @@ public interface FlintClient {
...
 
+  /**
+   * Create {@link FlintReader}.
+   *
+   * @param indexName index name.
+   * @param shardId shard id.
+   * @param query DSL query. DSL query is null means match_all
+   * @return {@link FlintReader}.
+   */
+  FlintReader createReader(String indexName, String shardId, String query);
```
Comment (dai-chen) [inline flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java:85]: Is `shardId` concept binding to `FlintOpenSearchClient` implementation or generic?

File: flint-spark-integration/src/main/scala/org/apache/spark/opensearch/table/OpenSearchTable.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Represents an OpenSearch table.
+ *
+ * @param tableName
+ *   The name of the table.
+ * @param metadata
+ *   Metadata of the table.
+ */
+case class OpenSearchTable(tableName: String, metadata: Map[String, FlintMetadata]) {
```
Comment (dai-chen) [inline flint-spark-integration/src/main/scala/org/apache/spark/opensearch/table/OpenSearchTable.scala:24]: Is using `FlintMetadata` temporary for only fetching index setting? Or is there hard dependency between OS table and Flint index in future?

File: flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java (hunk 1 above)
Comment (penghuo) [reply flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java:85]: bind to OpenSearch

File: flint-spark-integration/src/main/scala/org/apache/spark/opensearch/table/OpenSearchTable.scala (hunk 2 above)
Comment (penghuo) [reply flint-spark-integration/src/main/scala/org/apache/spark/opensearch/table/OpenSearchTable.scala:24]: only for index setting and mapping, not binding to real index

**Review by dai-chen — APPROVED** [review]

File: flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java (hunk 1 above)
Comment (dai-chen) [reply flint-core/src/main/scala/org/opensearch/flint/core/FlintClient.java:85]: We can abstract this task/split info later.

**Pull request merged**